File: game-dividead/dividead-jmedialayer/src/com/talestra/dividead/LZ.java
package com.talestra.dividead;


import com.jtransc.annotation.JTranscInline;
import com.jtransc.io.ra.RAByteArray;
import jmedialayer.util.FastMemByte;

import java.util.Arrays;
import java.util.Objects;

public class LZ {
    static public boolean isCompressed(byte[] data) {
        return Objects.equals(RAStreamExKt.readStringz(new RAByteArray(data), 2), "LZ");
    }

    static public int getUncompressedSize(byte[] data) {
        return new RAByteArray(data).sliceAvailable(6).readS32_LE();
    }

    static public byte[] uncompress(byte[] data) {
        if (data == null || data.length == 0) return new byte[0];
        RAByteArray sdata = new RAByteArray(data);
        String magic = RAStreamExKt.readStringz(sdata, 2);
        int compressedSize = sdata.readS32_LE();
        int uncompressedSize = sdata.readS32_LE();
        if (!Objects.equals(magic, "LZ")) throw new RuntimeException("Invalid LZ stream");
        return _decodeFast(sdata.readBytes((long) compressedSize), uncompressedSize);
    }

    static private byte[] _decodeFast(byte[] input, int uncompressedSize) {
        int ip = 0;
        int il = input.length;


        byte[] o2 = new byte[uncompressedSize + 0x1000];
        int op = 0x1000;
        int ringStart = 0xFEE;

        FastMemByte.selectSRC(input);
        FastMemByte.selectDST(o2);

        while (ip < il) {
            int code = FastMemByte.getSRC_u(ip++) | 0x100;

            while (code != 1) {
                // Uncompressed
                if ((code & 1) != 0) {
                    FastMemByte.setDST(op++, FastMemByte.getSRC_u(ip++));
                }
                // Compressed
                else {
                    if (ip >= il) break;
                    int paramL = FastMemByte.getSRC_u(ip++);
                    int paramH = FastMemByte.getSRC_u(ip++);
                    int param = paramL | (paramH << 8);
                    int ringOffset = extractPosition(param);
                    int ringLength = extractCount(param);
                    int convertedP2 = ((ringStart + op) & 0xFFF) - ringOffset;
                    int convertedP = (convertedP2 < 0) ? convertedP2 + 0x1000 : convertedP2;
                    int outputReadOffset = op - convertedP;

                    for (int n = 0; n < ringLength; n++) {
                        FastMemByte.setDST(op + n, FastMemByte.getDST_u(outputReadOffset + n));
                    }
                    op += ringLength;
                }

                code = code >>> 1;
            }
        }

        return Arrays.copyOfRange(o2, 0x1000, 0x1000 + uncompressedSize);
    }

    @JTranscInline
    static private int extractPosition(int param) {
        return (param & 0xFF) | ((param >>> 4) & 0xF00);
    }

    @JTranscInline
    static private int extractCount(int param) {
        return ((param >>> 8) & 0xF) + 3;
    }
}

File: dividead-jmedialayer/src/com/talestra/dividead/LZ.kt
package com.talestra.dividead

import com.jtransc.io.ra.RAByteArray
import jmedialayer.util.FastMemByte

/*
object LZ {
	@JvmStatic fun isCompressed(data: ByteArray): Boolean = RAByteArray(data).readStringz(2) == "LZ"

	@JvmStatic fun getUncompressedSize(data: ByteArray): Int = RAByteArray(data).sliceAvailable(6).readS32_LE()

	@JvmStatic fun uncompress(data: ByteArray): ByteArray {
		val sdata = RAByteArray(data)
		val magic = sdata.readStringz(2)
		val compressedSize = sdata.readS32_LE()
		val uncompressedSize = sdata.readS32_LE()
		if (magic != "LZ") throw RuntimeException("Invalid LZ stream")
		return _decodeFast(sdata.readBytes(compressedSize.toLong()), uncompressedSize)
	}

	@JvmStatic private fun _decodeFast(input: ByteArray, uncompressedSize: Int): ByteArray {
		var ip = 0
		val il = input.size


		val o2 = ByteArray(uncompressedSize + 0x1000)
		var op = 0x1000
		val ringStart = 0xFEE

		FastMemByte.selectSRC(input)
		FastMemByte.selectDST(o2)

		while (ip < il) {
			var code = FastMemByte.getSRC_u(ip++) or 0x100

			while (code != 1) {
				// Uncompressed
				if ((code and 1) != 0) {
					FastMemByte.setDST(op++, FastMemByte.getSRC_u(ip++))
				}
				// Compressed
				else {
					if (ip >= il) break
					val paramL = FastMemByte.getSRC_u(ip++)
					val paramH = FastMemByte.getSRC_u(ip++)
					val param = paramL or (paramH shl 8)
					val ringOffset = extractPosition(param)
					val ringLength = extractCount(param)
					val convertedP2 = ((ringStart + op) and 0xFFF) - ringOffset
					val convertedP = if (convertedP2 < 0) convertedP2 + 0x1000 else convertedP2
					val outputReadOffset = op - convertedP
					for (n in 0 until ringLength) {
						FastMemByte.setDST(op + n, FastMemByte.getDST_u(outputReadOffset + n))
					}
					op += ringLength
				}

				code = code ushr 1
			}
		}

		return o2.copyOfRange(0x1000, 0x1000 + uncompressedSize)
	}

	@JTranscInline @JvmStatic private fun extractPosition(param: Int): Int = (param and 0xFF) or ((param ushr 4) and 0xF00)
	@JTranscInline @JvmStatic private fun extractCount(param: Int): Int = ((param ushr 8) and 0xF) + 3
}
	*/
